crates/bevy_app/src/plugin.rs

```diff
@@ -56,19 +89,46 @@ use std::any::Any;
 /// # fn damp_flickering() {}
 /// ````
 pub trait Plugin: Downcast + Any + Send + Sync {
-    /// Configures the [`App`] to which this plugin is added.
-    fn build(&self, app: &mut App);
+    /// Returns required sub apps before finalizing this plugin.
+    fn require_sub_apps(&self) -> Vec<InternedAppLabel> {
+        Vec::new()
+    }
 
-    /// Has the plugin finished its setup? This can be useful for plugins that need something
-    /// asynchronous to happen before they can finish their setup, like the initialization of a renderer.
-    /// Once the plugin is ready, [`finish`](Plugin::finish) should be called.
-    fn ready(&self, _app: &App) -> bool {
+    /// Pre-configures the [`App`] to which this plugin is added. This is usually executed before
+    /// the event loop is started.
+    fn init(&self, _app: &mut App) {
+        // do nothing
+    }
+
+    /// Is the plugin ready to be built?
+    fn ready_to_build(&self, _app: &mut App) -> bool {
         true
     }
 
-    /// Finish adding this plugin to the [`App`], once all plugins registered are ready. This can
+    /// Builds the [`Plugin`] resources. This is usually executed inside an event loop.
+    fn build(&self, _app: &mut App) {
```

Comment:
If we end up wanting to change the name here, we should slap a `#[deprecated]` annotation on this method to ease migration.

Reply:
I re-added `build` => calls `init` and it's deprecated now.